src/loader/module_wrap.cc

```diff
@@ -0,0 +1,493 @@
+#include <algorithm>
+#include "module_wrap.h"
+
+#include "../env.h"
+#include "../node_url.h"
+#include "util.h"
+#include "util-inl.h"
+
+namespace node {
+namespace loader {
+using v8::Local;
+using v8::Persistent;
+using v8::MaybeLocal;
+using v8::Promise;
+using v8::Object;
+using v8::Module;
+using v8::Value;
+using v8::Isolate;
+using v8::String;
+using v8::TryCatch;
+using v8::FunctionCallbackInfo;
+using v8::Context;
+using v8::JSON;
+using v8::PropertyCallbackInfo;
+using v8::Function;
+using v8::IntegrityLevel;
+using v8::ScriptCompiler;
+using v8::ScriptOrigin;
+using v8::Integer;
+using v8::FunctionTemplate;
+using v8::EscapableHandleScope;
+using v8::PropertyDescriptor;
+
+const std::string EXTENSIONS[] = {".mjs", ".js", ".json", ".node"};
+std::map<int, std::vector<ModuleWrap*>*> ModuleWrap::module_map_;
+
+ModuleWrap::ModuleWrap(Environment* env,
+                      Local<Object> object,
+                      Local<Module> module,
+                      Local<String> url
+                     ) : BaseObject(env, object) {
+  Isolate* iso = Isolate::GetCurrent();
+  module_.Reset(iso, module);
+  url_.Reset(iso, url);
+}
+
+ModuleWrap::~ModuleWrap() {
+  auto module = module_.Get(Isolate::GetCurrent());
+  auto same_hash = module_map_[module->GetIdentityHash()];
+  auto it = std::find(same_hash->begin(), same_hash->end(), this);
+  if (it != same_hash->end()) {
+    same_hash->erase(it);
+  }
+  module_.Reset();
+}
+
+void ModuleWrap::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  Isolate* iso = args.GetIsolate();
+  if (!args.IsConstructCall()) {
+    env->ThrowError("constructor must be called using new");
+    return;
+  }
+  if (args.Length() != 2) {
+    env->ThrowError("constructor must have exactly 2 argument (string, string)");
+    return;
+  }
+
+  if (!args[0]->IsString()) {
+    env->ThrowError("first argument is not a string");
+    return;
+  }
+  auto arg = args[0]->ToString();
+  String::Utf8Value source_text(arg);
+
+  if (!args[1]->IsString()) {
+    env->ThrowError("second argument is not a string");
+    return;
+  }
+  Local<String> url = args[1]->ToString();
+
+  Local<Module> mod;
+  // compile
+  {
+    ScriptOrigin origin(url,
+                            Integer::New(iso, 0),
+                            Integer::New(iso, 0),
+                            False(iso),
+                            Integer::New(iso, 0),
+                            String::NewFromUtf8(iso, ""),
+                            False(iso),
+                            False(iso),
+                            True(iso));
+    ScriptCompiler::Source source(arg, origin);
+    {
+      TryCatch compileCatch(iso);
+      auto maybe_mod = ScriptCompiler::CompileModule(iso, &source);
+      if (compileCatch.HasCaught()) {
+        compileCatch.ReThrow();
```

Comment:
Not sure but I think all the `TryCatch` can be replaced by `maybe.isEmpty()` checks. /cc @addaleax